### PR TITLE
Documentation: constitutive models in the developer guide

### PR DESCRIPTION
**Pull request by CusiniM**
The branch is on `readTheDocs` if you prefer to review it there.

### CODE REVIEW

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
+.. _constitutiveModels:
```
Comment (CusiniM) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:1]: if anyone understands why sphinx says that this label is duplicated I d love some help...

```
../../../../../../workspace/cusini1/geosx/geosx_dev/GEOSX/src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:5: WARNING: duplicate label constitutivemodels, other instance in /usr/workspace/cusini1/geosx/geosx_dev/GEOSX/src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst

**Review by joshua-white — APPROVED** [review]
Looks great

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+################################################################################
+Constitutive models
+################################################################################
+
+In GEOSX, all constitutive models defining fluid and rock properties are implemented
+in the namespace ``constitutive`` and derive from a common base class,
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:8]:
```suggestion
in the namespace ``constitutive`` and derived from a common base class,
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  - relative permeability relationships,
+  - porosity and permeability dependencies on state variables,
+  - contact laws.
+
+Storage, allocation, and update of properties
+-----------------------------------------------------
+Each constitutive model owns, as member variables, ``LvArray::Array`` containers
+that hold the properties (or fields) and their derivatives with respect to the
+other fields needed to update each property. Each property is stored as an array with the
+first dimension representing the elementIndex and the second dimension the index of the
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:26]:
```suggestion
first dimension representing the elementIndex and the second dimension storing the index of the
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+.. literalinclude:: /coreComponents/constitutive/fluid/SingleFluidBase.cpp
+   :language: c++
+   :start-after: //START_SPHINX_INCLUDE_00
+   :end-before: //END_SPHINX_INCLUDE_00
+
+Any property or field stored on a constitutive model must be updated within a computational
+kernel to ensure that `host` and `device` memory in GPUs are properly synced, and that any
+updates are performed on `device`. Some properties are updated
+within finite element kernels of specific physics (such as stress in a mechanics kernel). Consequently,
+for each constitutive model class, a corresponding `nameOfTheModelUpdates` which, since it only contains
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:56]:
```suggestion
for each constitutive model class, a corresponding `nameOfTheModelUpdates`, which only contains
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+.. literalinclude:: /coreComponents/constitutive/fluid/SingleFluidBase.hpp
+   :language: c++
+   :start-after: //START_SPHINX_INCLUDE_02
+   :end-before: //END_SPHINX_INCLUDE_02
+
+Compound models
+========================================================
+
+Compound constitutive models are employed to mimic the behavior of a material that
+requires a combination of constitutive models linked together.  These compound models
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:77]:
```suggestion
requires a combination of constitutive models linked together. These compound models
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+first dimension representing the elementIndex and the second dimension the index of the
+integration point. These dimensions are determined by the number of elements of the
+subregion on which each constitutive model is registered, and by the chosen discretization
+method. Vector and tensor fields have an additional dimension to identify
+their components. Similarly, an additional dimension is
+necessary for multiphase fluid models with properties defined for each component in each phase.
+For example, a single-phase fluid model where density and viscosity are
+functions of the fluid pressure has the following members:
+
+.. literalinclude:: /coreComponents/constitutive/fluid/SingleFluidBase.hpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:35]:
```suggestion
.. literalinclude:: ../fluid/SingleFluidBase.hpp
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+   :end-before: //END_SPHINX_INCLUDE_00
+
+Resizing all fields of the constitutive models happens during the initialization phase by
+the ``ConstitutiveManger`` through a call to ``ConstitutiveManger::hangConstitutiveRelation``,
+which sets the appropriate subRegion as the parent Group of each constitutive model object.
+This function also resizes all fields based on the size of the subregion and the number of quadrature
+points on it, by calling ``CONSTITUTIVE_MODEL::allocateConstitutiveData``. For the
+single phase fluid example used before, this call is:
+
+.. literalinclude:: /coreComponents/constitutive/fluid/SingleFluidBase.cpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:47]:
```suggestion
.. literalinclude:: ../fluid/SingleFluidBase.cpp
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+Any property or field stored on a constitutive model must be updated within a computational
+kernel to ensure that `host` and `device` memory in GPUs are properly synced, and that any
+updates are performed on `device`. Some properties are updated
+within finite element kernels of specific physics (such as stress in a mechanics kernel). Consequently,
+for each constitutive model class, a corresponding `nameOfTheModelUpdates` which, since it only contains
+``LvArray::arrayView`` containers to the data, can be captured by value inside computational kernels.
+For example, for the single phase fluid model `Updates` are:
+
+.. literalinclude:: /coreComponents/constitutive/fluid/SingleFluidBase.hpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:60]:
```suggestion
.. literalinclude:: ../fluid/SingleFluidBase.hpp
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+   :language: c++
+   :start-after: //START_SPHINX_INCLUDE_01
+   :end-before: //END_SPHINX_INCLUDE_01
+
+Because `Updates` classes are responsible for updating the fields owned by the constitutive models,
+they also implement all functions needed to perform property updates, such as:
+
+.. literalinclude:: /coreComponents/constitutive/fluid/SingleFluidBase.hpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:68]:
```suggestion
.. literalinclude:: ../fluid/SingleFluidBase.hpp
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+---------------------------------------------------------
+
+``CoupledSolid`` models are employed to represent porous materials that require
+both a mechanical behavior and constitutive laws that describe the
+dependency of porosity and permeability on the primary unknowns.
+
+The base class ``CoupledSolidBase`` implements some basic behaviors
+and is used to access a generic ``CoupledSolid`` in a physics solver:
+
+.. literalinclude:: /coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:91]:
```suggestion
.. literalinclude:: ../../physicsSolvers/fluidFlow/SinglePhaseBase.cpp
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+   :language: c++
+   :start-after: //START_SPHINX_INCLUDE_COUPLEDSOLID
+   :end-before: //END_SPHINX_INCLUDE_COUPLEDSOLID
+
+Additionally, a `template class` defines a base ``CoupledSolid`` model
+templated on the types of solid, porosity, and permeability models:
+
+.. literalinclude:: /coreComponents/constitutive/solid/CoupledSolid.hpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:99]:
```suggestion
.. literalinclude:: ../solid/CoupledSolid.hpp
```

File: src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+   :language: c++
+   :start-after: //START_SPHINX_INCLUDE_00
+   :end-before: //END_SPHINX_INCLUDE_00
+
+While physics solvers that need a porous material only interface with a compound model,
+this one has access to the standalone models needed:
+
+.. literalinclude:: /coreComponents/constitutive/solid/CoupledSolid.hpp
```
Comment (jhuang2601) [inline src/coreComponents/constitutive/docs/constitutiveDeveloperGuide.rst:107]:
```suggestion
.. literalinclude:: ../solid/CoupledSolid.hpp
```

**Review by jhuang2601 — APPROVED** [review]
Great job!

**Pull request merged**